monai/data/image_writer.py

```diff
@@ -0,0 +1,395 @@
+    """
+    Write data and metadata into files on disk using ITK-python.
+    """
+
+    def __init__(self, output_dtype: DtypeLike = np.float32, **kwargs):
+        super().__init__(output_dtype=output_dtype, affine=None, channel_dim=0, **kwargs)
+
+    def set_data_array(self, data_array, channel_dim: Optional[int] = 0, squeeze_end_dims: bool = True, **kwargs):
+        """Convert ``data_array`` into 'channel-last' numpy ndarray."""
+        self.data_obj = self.convert_to_channel_last(
+            data=data_array,
+            channel_dim=channel_dim,
+            squeeze_end_dims=squeeze_end_dims,
+            spatial_ndim=kwargs.pop("spatial_ndim", 3),
+            contiguous=kwargs.pop("contiguous", True),
+        )
+        self.channel_dim = channel_dim
+
+    def set_metadata(self, meta_dict: Optional[Mapping] = None, resample: bool = True, **options):
+        """Resample ``self.dataobj`` if needed.  This method assumes ``self.data_obj`` is a 'channel-last' ndarray."""
+        original_affine, affine, spatial_shape = self.get_meta_info(meta_dict)
+        self.data_obj, self.affine = self.resample_if_needed(
+            data_array=self.data_obj,
+            affine=affine,
+            target_affine=original_affine if resample else None,
+            output_spatial_shape=spatial_shape,
+            mode=options.pop("mode", GridSampleMode.BILINEAR),
+            padding_mode=options.pop("padding_mode", GridSamplePadMode.BORDER),
+            align_corners=options.pop("align_corners", False),
+            dtype=options.pop("dtype", np.float64),
+        )
+
+    def write(self, filename_or_obj: PathLike, verbose: bool = False, **kwargs):
+        """Create an ITK object from ``self.data_obj`` and call ``itk.imwrite``"""
+        super().write(filename_or_obj, verbose=verbose)
+        self.data_obj = self.create_backend_obj(
+            self.data_obj, channel_dim=self.channel_dim, affine=self.affine, dtype=self.output_dtype, **kwargs  # type: ignore
+        )
+        itk.imwrite(
+            self.data_obj,
+            filename_or_obj,
+            compression=kwargs.pop("compression", False),
+            imageio=kwargs.pop("imageio", None),
+        )
+
+    @classmethod
+    def create_backend_obj(
+        cls,
+        data_array: NdarrayOrTensor,
+        channel_dim: Optional[int] = 0,
+        affine: Optional[NdarrayOrTensor] = None,
+        dtype: DtypeLike = np.float32,
+        **kwargs,
+    ):
+        """Create an ITK object from ``data_array``.  This method assumes a 'channel-last' ``data_array``."""
+        data_array = super().create_backend_obj(data_array)
+        _is_vec = channel_dim is not None
+        if _is_vec:
+            data_array = np.moveaxis(data_array, -1, 0)  # from channel last to channel first
+        data_array = data_array.T.astype(dtype, copy=True, order="C")
+        itk_obj = itk.GetImageFromArray(data_array, is_vector=_is_vec, ttype=kwargs.pop("ttype", None))
```

Comment:
May I know the usage case of this `is_vector` arg here?
I don't quite understand the relationship between `is_vec` and ITK dim order vs numpy dim order.

Thanks in advance.

Reply:
this flag and the dtype will determine the underlying itk object type, for example
```py
>>> import numpy as np
>>> array = np.zeros((1,2,3))
>>> import itk
>>> itk.GetImageFromArray(array, is_vector=False)
<itk.itkImagePython.itkImageD3; proxy of <Swig Object of type 'itkImageD3 *' at 0x7f8b5aedb600> >
>>> itk.GetImageFromArray(array, is_vector=True)
<itk.itkVectorImagePython.itkVectorImageD2; proxy of <Swig Object of type 'itkVectorImageD2 *' at 0x7f8b68fad1e0> >
```
the same array could be interpreted as a 3d spatial image or a 2d vector image.

Reply:
I still don't understand this line: `_is_vec = channel_dim is not None`, why if `channel_dim` is not None, then the data is **2d vector image**?

Thanks.

Reply:
in the example `np.zeros((1,2,3))` could be saved as a 2d image with channels (e.g. `channel_dim=-1` for channel-last) or a 3d image without any channel (`channel_dim=None`).

Reply:
But if the data is 4d like `[128, 128, 128, 3]` and `channel_dim` is `-1`, it will be treated as `_is_vec = True`?

Thanks.